DataStructures/Graphs/Ford_Fulkerson_Max_Flow.java

```diff
@@ -0,0 +1,137 @@
+            graph[graphData.get(i)].add(arc);
+            edges.put(arc, graphData.get(i + 2));
+        }
+    }
+
+    private static boolean bfsPathFinder(int from, int to, HashSet<String>[] graph, TreeMap<String, Integer> edges, ArrayList<String> path) {
+        boolean[] visited = new boolean[graphNodes];
+        Queue<Integer> queue = new LinkedList<>();
+        int[] parent = new int[graphNodes];
+        visited[from] = true;
+        queue.add(from);
+        while (!queue.isEmpty() && !visited[to]) {
+            int ini = queue.poll();
+            for (String edge: graph[ini]) {
+                int fin = Integer.parseInt(edge.substring(edge.indexOf('-') + 1));
+                //if the node has not been visited yet AND the edge connecting them has still capacity available
+                if (!visited[fin] && edges.get(edge) > 0) {
+                    parent[fin] = ini;
+                    visited[fin] = true;
+                    queue.add(fin);
+                }
+            }
+        }
+        //once a path has been found, the algorithm saves all the edges to update their flow back in fordFulkerson method
+        if (visited[to]) {
+            int ini = to;
+            path.clear();
+            while (ini != from) {
+                int aux = parent[ini];
+                path.add(0, aux + "-" + ini);
+                ini = aux;
+            }
+        }
+        return visited[to];
+    }
+
+    private static int fordFulkerson (HashSet<String>[] graph, TreeMap<String, Integer> edges, int source, int sink) {
+        int maxFlow = 0;
+        ArrayList<String> path = new ArrayList<>();
+        //while there is a path between source and sink with capacity available, keep flow stream going
+        while (bfsPathFinder(source, sink, graph, edges, path)) {
+            int[] pathFlow = {Integer.MAX_VALUE};
+            //int as an array to use lambda for loop
+            path.forEach(edge -> pathFlow[0] = Math.min(pathFlow[0], edges.get(edge)));
+            //by consulting the capacity available of every edge used in this new path,
+            //bottleneck is calculated as the minimum capacity of the path. It corresponds
+            //to the number in which the flow can be incremented
+            path.forEach(edge -> edges.put(edge, edges.get(edge) - pathFlow[0]));
+            //each used edge gets its capacity reduced by the path bottleneck value
+            maxFlow += pathFlow[0];
+        }
+        return maxFlow;
+    }
+
+    private static ArrayList<Integer> pickExample(int choice) {
+        //from, to, capacity, null (to visually separate edges)
+        switch (choice) {
+            case 0: {
+                //From 0 (source) to 10 (sink) should return 7
+                source = 0;
+                sink = 10;
+                graphNodes = 11;
+                return new ArrayList<>(Arrays.asList(0, 1, 7, null, 0, 2, 2, null, 0, 3, 1, null, 1, 4, 2,
```

Comment:
Why do you add a null value in the `i+3` position?

Reply:
Because with that null value, the code is easier to read.
Graph edges are set in groups of three values, from, to and weight. 
Null values separate those edges, and programmers don't lose time trying to understand how many edges are and which values conform them.